tests/queries.ts

```diff
@@ -136,3 +136,21 @@ testClient(async function multiQueryWithManyQueryTypeArray() {
 
   assertEquals(result[2].rows.length, 2);
 });
+
+testClient(async function transactionWithConcurrentQueries() {
+  const result = await CLIENT.query("BEGIN");
+
+  assertEquals(result.rows.length, 0);
+  const concurrentCount = 5;
+  const queries = [...Array(concurrentCount)].map((_, i) => {
+    return CLIENT.query({
```

Comment:
The "bug" is there is no await here (so it reuses the client in an unexpected way).

I think making client private and the query api going through pool is the right play (this is a good test case - it certainly shouldn't error with a confusing low-level error).

Reply:
I think if the query API goes through the pool it may not allow for transactions to reuse the same client?

> The "bug" is there is no await here (so it reuses the client in an unexpected way).

The intent was to perform the queries concurrently rather than serially. Adding an await here doesn't change the behavior, since `Array.map` doesn't  await Promises. In both cases, `Promise.all` will receive an array of promises that can be fulfilled in any order.

The query queuing behavior is supported in `node-postgres`, which is where the expectation came from. They do a more complicated query pulse which was the inspiration for this.

https://github.com/brianc/node-postgres/blob/master/packages/pg/lib/client.js#L478

I do see value is firing queries that can happen in any order, say fetching or inserting several unrelated things within a transaction.

Reply:
Good point. You're right even the await doesnt fix it.

But can we reuse the DeferredQueue machinery to limit it to one execute/query at a time on a Client?

Reply:
I did briefly look into it but the semantics were different. It's worth further investigation.